### PR TITLE
Rust automated contract tests implementation

### PR DESCRIPTION
**Pull request by mariopil**
**Referenced issue**

If applies, the issue this PR belongs to.

**Summary of changes**

First try on Rust automated tests implementation. Unfortunately there are issues with parsing transaction result meta XDR. The problem is with `stellar-xdr` library, which only supports `TransactionMetaV1` and `V2`, but futurenet returns `V3`. Currently I don't have solution for this...

**Reviewer recommendations**

How to make the review easier. i.e "Its a big PR and the recommendation is to go through each commit" 
or "In order to run the tests, execute ..."

### CODE REVIEW

**Comment by eloylp** [comment]
>  The problem is with stellar-xdr library, which only supports TransactionMetaV1 and V2, but futurenet returns V3

Hello ! I was trying to decode the transaction meta with the CLI for [this transaction](https://laboratory.stellar.org/#xdr-viewer?input=AAAAAwAAAAAAAAACAAAAAwAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj%2F5wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAQAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj%2F5wAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAMAAAAAAAAB9QAAAABkuvKlAAAAAAAAAAEAAAADAAAAAwAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj%2F5wAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAMAAAAAAAAB9QAAAABkuvKlAAAAAAAAAAEAAAH1AAAAAAAAAAAG0ZX%2FOzai1O6AQpGE0XNYqFaYVdHnYaNktZQr6%2FHBsgAjhvJvwP%2BcAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAADAAAAAAAAAfUAAAAAZLrypQAAAAAAAAAAAAAB9QAAAAAAAAAADvIyr2G%2Bt9EXKjrTpfdOSF8aXW8xLMPEm8A7Xa2KejsNvS%2FBN6MAAAAAAfUAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAA&type=TransactionMeta):

```bash
soroban lab xdr dec --type TransactionMeta --xdr "AAAAAwAAAAAAAAACAAAAAwAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj/5wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAQAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj/5wAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAMAAAAAAAAB9QAAAABkuvKlAAAAAAAAAAEAAAADAAAAAwAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj/5wAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAMAAAAAAAAB9QAAAABkuvKlAAAAAAAAAAEAAAH1AAAAAAAAAAAG0ZX/Ozai1O6AQpGE0XNYqFaYVdHnYaNktZQr6/HBsgAjhvJvwP+cAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAADAAAAAAAAAfUAAAAAZLrypQAAAAAAAAAAAAAB9QAAAAAAAAAADvIyr2G+t9EXKjrTpfdOSF8aXW8xLMPEm8A7Xa2KejsNvS/BN6MAAAAAAfUAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAA"
```

It looks like CLI is using `TransactionMeta` type for decoding , without any version suffix. The CLI (which is Rust) is able to parse the base64, which looks like is `V3` (got it from laboratory):

image

Its curious, because if i try to parse by specifying `TransactionMetaV3` explicitly,  got an error:

```bash
soroban lab xdr dec --type TransactionMetaV3 --xdr "AAAAAwAAAAAAAAACAAAAAwAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj/5wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAQAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj/5wAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAMAAAAAAAAB9QAAAABkuvKlAAAAAAAAAAEAAAADAAAAAwAAAfUAAAAAAAAAAAbRlf87NqLU7oBCkYTRc1ioVphV0edho2S1lCvr8cGyDeC2s6dj/5wAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAMAAAAAAAAB9QAAAABkuvKlAAAAAAAAAAEAAAH1AAAAAAAAAAAG0ZX/Ozai1O6AQpGE0XNYqFaYVdHnYaNktZQr6/HBsgAjhvJvwP+cAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAADAAAAAAAAAfUAAAAAZLrypQAAAAAAAAAAAAAB9QAAAAAAAAAADvIyr2G+t9EXKjrTpfdOSF8aXW8xLMPEm8A7Xa2KejsNvS/BN6MAAAAAAfUAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAA"
error: decode: parsing xdr: xdr value invalid
```

Looks like there is an enum (not struct) called [TransactionMeta](https://docs.rs/stellar-xdr/0.0.17/stellar_xdr/next/enum.TransactionMeta.html) which looks like could have the logic for automatically read a header and use the proper struct under the hood ([See code here](https://github.com/stellar/rs-stellar-xdr/blob/main/src/next/generated.rs#L20099-L20116)). I would try with that type , as seems is the same used by CLI, which is able to parse it :)

**Comment by mariopil** [comment]
Indeed it works with a client, but doesn't work from the lib...

**Comment by mariopil** [comment]
It's possible to generate output in json format so we could use the client to decode the xdr and the parse it :)

**Comment by mariopil** [comment]
I've updated the Makefile.toml scripts. Transaction XDR results are parsed using the soroban client. For parsing the transactions and decoded XDRs js scripts are used (it's just very easy to parse jsons in js). I think the rust test files can be removed, Makefiles do the job good :)

**Comment by mariopil** [comment]
PR is now ready to review. I've completely removed the Rust tests and added 3 makefiles that can be launched via `cargo make` (one for each contract). There's also a script that runs all tests, `integration_tests.sh` which is launched by github job. Currently the exit code of the this script is always 0 as the Raffle test is not working because of some futurenet issues.

File: contracts/raffle/Makefile.toml [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  
+  curl "${friendbot}?addr=$(soroban config identity address admin)"
+'''
+
+[tasks.create_account_players]
+script = '''
+  for PLAYER in 1 2 3
+  do
+    soroban config identity generate player_$PLAYER 2>/tmp/raffle_error
+    ERR=`cat /tmp/raffle_error`
```
Comment (mariopil) [inline contracts/raffle/Makefile.toml:44]: There is a nicer way to store `stderr` in a variable in bash, but for some reason that was the only way that worked when run in a github job.

**Review by geofmureithi — APPROVED** [review]
LGTM

**Review by eloylp — APPROVED** [review]
I think this would be good for now (we have nothing). But certainly i had a lot of hopes we would be able to create tests cases from Rust code :broken_heart: . I understand we are currently limited.

Okay, lets re-check this in the future as things evolve and stabilize ! :)

**Pull request merged**